Draft: The MEG [ Test rework, changing api code layout, implement ClassifyElement, implement JoinElement, Add Overseer]

CollinValley opened the pull request:
I know this is kind of a mega commit stack, I just wanted you all to take a look at it before we try to open up individual PRs.  This solves quite a few of our open issue, and I think after this is merged we can start trying to write elements, and form there create a functioning router. Which is pretty cool.  As far as I am aware, this code works as written. We could use more tests as always, but it is a good start.

CollinValley committed:
Implement ClassifyElement

This commit implements the ClassifyElement, a generic Asynchronous
element that takes in a packet, and sends the packet out of a port
based on a user provided classify function. In this basic implementation,
there a few less than ideal cases.

The classify function, right now, outputs a tuple with a number
determining what port to send the packet out of, perhaps this could be
a different mechanism if we so desire.

The Consumer, which takes in a packet and forwards it, via a channel,
to the appropriate provider, will stop processing packets if any one of
the channels to a port becomes full. This may be the implementation
that we want moving foward, but we should discuss this. Additionally,
at the beginning of each poll we check all the channels to see if any
are full, and await the producer of the full task.  Iterating over every
channel on each poll is not very efficient, perhaps we could do a full
buffer check at the tail, when we push a packet, and sleep then
instead, and only that the appropriate channel is less full on wake,
saving us some cycles. However, if the number of ports is low, this
cost may not be a big deal, and we can leave the function as is.

CollinValley committed:
join_element.rs: Implement JoinElement

This join element works as essentially the mirror to the classify
element. It has many consumers, each with a channel on which they
can place packets to be pulled by the single provider. This
implementation also requires an overseer component, and it's drop
logic is somewhat different. It keeps track of which consumers are
counted as dead by substituting a None for ther wake_consumer
Receiver<task::Task>. This is done to ensure that a counted dead
consumer is not double counted. The overseer iterates over each
packet_channel, and will attempt to wake the provider if it detects
that any of the channels contain a packet.

Issues: This join is unfair, since the provider always starts with the
first consumer, rather than executing a try round robin. This will be
resolved before we merge this element.

CollinValley committed:
Element: Move api::element to its own module

This commt moves the Element, ElementLink and it's related tests into
a new file, element.rs, that lives under the api directory.  This is better
for clarity in the crate, and help prevent merge conflicts.

CollinValley committed:
AsyncElement: Move api::async_elment to a module

This commit moves AsyncElement,  AsyncElementLink and its related
tests into a new file, async_element.rs, that lives under the api
directory.

CollinValley committed:
AsyncElement: Test cleanup

This commit removes a bunch of extranous tests from the AsyncElement
test suite. The remaining tests now all serve a unique purpose, and are
renamed to describe what they are testing, rather than how they are
testing. In general, this change also means that all tests are now
"collected", and assert that the correct packets reach their
destination.

CollinValley committed:
AsyncElement: Add overseer

Due to a race condition in the AsyncElementLink, the provider and
consumer sides can sometimes both enter sleep at the same time,
where each one thinks the channel is empty and full, respectivly, and
neither do work. This is a serious problem with our current architecture,
and will need to be resolved in the future. As a stop-gap, we are paying
a performance hit to run a third actor in the aynchronous links, the
overseer. The overseer checks whether the channel contains packets, and
attempts to awaken the appropriate side given the outcome. This
removes the race condition and makes the tests pass. When the
overseer detects that the packet channel is empty, and the provider has
dropped, it returns Ready.  Prior to that it always returns notReady on
a self scheduling loop.

CollinValley committed:
ClassifyElement: Add Overseer

This commit adds the overseer. It solves the same problem as the
previous commit which adds the overseer to the AsyncElement. In
this case it works slightly differently, since we need to keep
track of the number of providers that are still alive, so that
the overseer does not return until all of the provider channels
are empty, and the consumer has dropped. Then it is safe for the
overseer to cease operation. This is a bit roundabout, but without
the explicit check in crossbeam_channels for disconnection, this is
a fine proxy.

scgruber commented:
This is probably also a good opportunity to drop everything down a level so we can have multiple crates in the repo.

CollinValley committed:
join_element: Make provider pull fairly

In the previous implementation, the provider of the join element would
pull starting from the 1st consumer onward, until it gets a packet.
This leads to unfairness,  to solve this, we have the provider rememebr
which consumer it last received the packet from, and polls the consumer
immediatly after it first.

CollinValley commented:
Moving the api to it's own crate seems like a good idea.  I can do that before pushing the official PRs.

CollinValley commented:
Heads up: Resolved the Join element fairness issue this morning, commit is now in stack.

ddowl commented:
We can get those first 2 commits, moving the `Element` definitions to their own modules (`086f708`, `0930c69`), into a PR straight away and get them merged. Those lgtm.

ddowl commented:
I'm glad we're cleaning up the tests to be assertion-only. We should make a similar test cleanup commit for the synchronous element as well.

ddowl commented on src/api/classify_element.rs:
```diff
@@ -0,0 +1,349 @@
+    type Item = ();
+    type Error = ();
+
+    /// Same logic as AsyncElementConsumer, except if any of the channels are full we
```
We can do this in a separate PR, but if we're sharing similar logic with `AsyncElementConsumer`, we should be thinking about what parts are similar and if we can compose any `AsyncElement` components here to reduce the complexity.

CollinValley replied:
https://github.com/route-rs/route-rs/issues/29 Created issue.

ddowl commented on src/api/classify_element.rs:
```diff
@@ -0,0 +1,349 @@
+    }
+}
+
+/// Split Element Provider, exactly the same as AsyncElementProvider, but
```
`SplitElementProvider`?

CollinValley replied:
Oh needs cleanup, `ClassifyElement` used to be `SplitElement`

ddowl replied:
Can we not modify `AsyncElementProvider` to have the bounds `E: Element + ClassifyElement`? I guess then the bound isn't tight enough to be useful.

CollinValley replied:
Maybe,  I just didn't want to muck around with it at the time.  It seems possible. Though Classify doesn't do a transformation,  hence the different trait.  Classify implements `fn classify` which outputs essentially a port number with a packet.

ddowl replied:
Maybe @moosingin3space can weigh in?

CollinValley replied:
https://github.com/route-rs/route-rs/issues/32

ddowl commented on src/api/classify_element.rs:
```diff
@@ -0,0 +1,349 @@
+                if let Err(_) = self.await_consumer.try_send(task) {
+                    task::current().notify();
+                }
+                Ok(Async::NotReady)
```
This logic is repeated on line 102. We should decompose it into a function.

CollinValley replied:
Makes sense.

CollinValley replied:
https://github.com/route-rs/route-rs/issues/29 Created issue

ddowl commented on src/api/classify_element.rs:
```diff
@@ -0,0 +1,349 @@
+                }
+                Ok(Async::Ready(Some(packet)))
+            },
+            Ok(None) => {
```
It's a shame we can't use `try_ready!` here since we need the custom logic for the `Err(TryRecvError::Empty)` arm.

ddowl commented on src/api/classify_element.rs:
```diff
@@ -0,0 +1,349 @@
+    use futures::future::lazy;
+
+    #[allow(dead_code)]
+    struct IdentityElement {
```
IdentityElement should definitely be defined elsewhere for testing purposes so we don't have to keep redefining it. We can do that in a separate commit.

CollinValley replied:
Yeah I can open an issue

CollinValley replied:
https://github.com/route-rs/route-rs/issues/30

ddowl commented on src/api/classify_element.rs:
```diff
@@ -0,0 +1,349 @@
+        let default_channel_size = 10;
+        let number_branches = 2;
+        let packets = vec![0, 1, 2, 420, 1337, 3, 4, 5, 6, 7, 8, 9, 11];
+        let packet_generator = PacketIntervalGenerator::new(time::Duration::from_millis(100), packets.clone().into_iter());
```
Again for another PR, but I think we should prefer using `immediate_stream` for packet generators to keep tests speedy, and include a test for each component that confirms it can handle pauses in the input stream.

CollinValley replied:
:+1:

CollinValley replied:
https://github.com/route-rs/route-rs/issues/31

ddowl commented:
Other than a few nitpicks I really like the `ClassifyElement`. It's implementation is _mostly_ straightforward, save for task waking issues (DAMN YOU CROSSBEAM!). The tests show a simple example, but it's descriptive and cool to see it actually work.

That being said, let's get these commits out into different PRs so we can address each of the issues in isolation.

ddowl commented:
Closed since we've reviewed these changes in the meetup today.
